# Bienvenidos al curso de R básico con CCOL 
Impartido por Javier Fernández <br>
Conecta conmigo en mi Linkedin: https://www.linkedin.com/in/qfbtjavierfernandezsegura/<br>
Síguenos en nuestras redes sociales: https://www.facebook.com/CCOLMX/ 

En este curso abordaremos los principios básicos del lenguaje de programación R, el cual es un proyecto open source (gratuito y extensible) que se especializa en computo estadístico y la elaboración de gráficas. 

El lenguaje R es muy fácil de aprender y utilizar, así que al final de éstas sesiones, tengo la esperanza de que tengas una noción mucho más clara del lenguaje para que puedas leerlo y escribirlo e incluso comiences a utilizarlo en tu trabajo, escuela o proyectos de investigación. 

Iniciemos por lo más básico, ¿vale? <br>
Empecemos por instalar R y RStudio en tu computadora, esto lo podrás hacer a través de estos url<br> R:   https://cran.r-project.org/ <br> Rstudio: https://rstudio.com/products/rstudio/download/#download <br>
Un tutorial de cómo instalarlos podrás encontrarlo aquí https://medium.com/datos-y-ciencia/tutorial-instalaci%C3%B3n-r-y-rstudio-563771a29289 

Perfecto! Una vez instalado abran RStudio, den click en Archivo -> Nuevo -> Script y comencemos a jugar con R.

### Aritmética con R

R puede ser utilizado como una calculadora, solo basta con ingresar los números y los operadores de suma (+), resta (-), multiplicación (*), división (/) y presionar ctrl (o cmd en mac) seguido de la tecla enter para ejecutar la operación

In [1]:
# Probemos unas operaciones sencillas
3 + 4
6 + 12
5 + 5 
5 - 5 
3 * 5
(5 + 5) / 2 

[1] 7

[1] 18

[1] 10

[1] 0

[1] 15

[1] 5

También podemos hacer operaciones un poco más elaboradas como calcular la raíz cuadrada, hacer operaciones con pi, potenciar un número con otro, calcular el exponencial, el logaritmo o incluso convertir un valor en absoluto con los siguientes operadores

In [2]:
# Raíz cuadrada
sqrt(540)

# Exponencial
exp(5)

# Potenciación
5 ^ 10

# pi
pi

# logaritmo
log(pi)

# valores absolutos
abs(-33.5)


[1] 23.2379

[1] 148.4132

[1] 9765625

[1] 3.141593

[1] 1.14473

[1] 33.5

Ahora sabemos como utilizar R como una calculadora, escribiendo las operaciones y ejecutandolas para imprimirlas en la consola de Rstudio.

Muchas veces durante su trabajo con programación van a necesitar el guardar algunos de los resultados de esas operaciones en valores que se les conoce como **variables**.

Estas variables te permiten guardar valores u objetos para que puedas acceder a ellos fácilmente después sin tener que ejecutar la operación nuevamente.

Hagamos la prueba con un el área de un círculo, como sabremos de nuestras clases de geometría el área se calcula como Pi x radio al cuadrado.  Así que calculemos el area de un circulo con radio de 5 y guardemos ese valor en una variable que se llame **adc** por área del círculo.

Para hacer esto ponemos el nombre de la variable seguido del simbolo  <-  y después escribimos nuestra operación.
Este símbolo se puede colocar fácilmente con alt + - ó con los simbolos < y - uno despues del otro.

Después para acceder al valor guardado en nuestra variable basta con poner el nombre de nuestra variable et voilà! el valor se imprimirá en nuestra consola tras ejecutar.

In [3]:
adc  <-  pi * (5^2)
adc

[1] 78.53982

Para poner en práctica estos principios básicos quiero que hagas lo siguiente:

- Asigna el valor 5 a la variable **manzanas** e imprímela
- Asigna el valor 6 a la variable **naranjas** e imprímela

Además de sólo guardar y mostrar los valores en tus variables, puedes hacer operaciones aritméticas con la variables, prueba sumando manzanas y naranjas.

Ahora asigna esa operación a la variable **frutas** e imprimelas. <br>

Excelente, ahora que has aprendido sobre como hacer aritmética con R quiero que aprendamos sobre los tipos de datos básicos que tiene R.

### Tipos de datos básicos

R puede trabajar con muchos tipos de datos los básicos y más utilizados son:<br>

- Numeric 

y estos a su vez se dividen en floats (decimales) e integers (enteros)

- Boolean

Estos se les conoce como valores lógicos y son TRUE o FALSE

- Character

Estos son texto o cadenas de texto

   

In [4]:
a  <-  43

b <- "Cinco"

c <- TRUE

d <- FALSE

Si quisieramos saber qué tipo de valor está guardado en una variable, podemos hacer uso de la función class(Nombre de la variable). Por ejemplo:


In [5]:
class(a)
class(b)
class(c)
class(d)

[1] "numeric"

[1] "character"

[1] "logical"

[1] "logical"

Si quisieramos hacer operaciones con ellos por ejemplo **a + b** osea 43 más cinco

In [6]:
a + b

ERROR: Error in a + b: argumento no-numérico para operador binario


R nos va a marcar un error ya que no son del mismo tipo de dato. Pero si por ejemplo ejecutamos 

In [7]:
c + d

[1] 1

Nos regresa un 1 ya que nos permite ejecutar la operación booleana sin problema alguno. <br>

En más de una ocasión vamos a necesitar acomodar una serie de números o caracteres uno seguido del otro. Para eso utilizamos objectos que en R se les conoce como **vectores**

### Vectores

Para iniciar un vector la sintaxis es la siguiente: <br>

c(1,2,3,4) <br>

En caso de usar caracteres necesitaras encerrarlos entre comillas dobles<br>

c("a","b","c","d") <br>

Intenta creando un vectores con los siguientes valores: <br>

Un vector llamado **numericos** con los valores 7,8,9 y 10 <br>
Un vector llamado **caracteres** con las letras CCOL <br>
Un vector llamado **booleanos** con los valores TRUE, FALSE,TRUE,TRUE <br>
    

Ahora hagamos un ejercicio con vectores. 
Imagina que vas a un casino una semana de lunes a viernes y que jugaste poker y ruleta. Decidiste colectar lo que ganaste y perdiste en los 5 dias.

En el poker: <br>

- El Lunes ganaste 140 
- El Martes perdiste 50
- El Miércoles ganaste 20
- EL Jueves perdiste 120
- El Viernes ganaste 240

En la ruleta: <br>

- El Lunes perdiste 24
- El Martes perdiste 50
- El Miércoles ganaste 100
- El Jueves perdiste 350
- El Viernes ganaste 10

Para poder utilizar estos valores crea 2 vectores uno que se llame **ruleta** y otro que se llame **poker**

In [8]:
poker <- c(140, -50, 20, -120, 240)

ruleta <- c(-24,-50,100,-350,10)

In [9]:
poker
ruleta

[1]  140  -50   20 -120  240

[1]  -24  -50  100 -350   10

Ahora crea un vector llamado **dias** con los días de la semana en que jugaste

In [10]:
dias <- c("Lunes","Martes","Miércoles","Jueves","Viernes")

Como una lista de número no nos dice mucho, empecemos por rotular estos número con la función `names(Nombre del objeto)` podemos extraer el encabezado de un vector

In [11]:
names(poker)
names(ruleta)
names(dias)

NULL

NULL

NULL

Como podrás los vectores aun no tienen encabezados, por lo que podemos añadirselos, asignandole a names(Objecto) un vector con el titulo de cada valor, en este caso el vector **dias** que ya creamos. 

In [12]:
names(poker)  <- dias
names(ruleta) <- dias    

In [13]:
names(poker)
names(ruleta)

[1] "Lunes"     "Martes"    "Miércoles" "Jueves"    "Viernes"

[1] "Lunes"     "Martes"    "Miércoles" "Jueves"    "Viernes"

Ahora veamos como luce nuestros vectores **poker** y **ruleta**

In [14]:
poker
ruleta

Lunes    Martes Miércoles    Jueves   Viernes 
      140       -50        20      -120       240

Lunes    Martes Miércoles    Jueves   Viernes 
      -24       -50       100      -350        10

Otra cosa interesante de los vectores es que, siempre y cuando sean numéricos, podemos hacer operaciones con ellos, veamos

In [15]:
poker + ruleta

Lunes    Martes Miércoles    Jueves   Viernes 
      116      -100       120      -470       250

Como vemos podemos apreciar cuanto es lo que ganamos y perdimos en cada días. Y ¿si quisieramos hacer la sumatoria de cada juego por separado? podríamos hacer lo siguiente: <br>
Ver el vector poker

In [16]:
poker

Lunes    Martes Miércoles    Jueves   Viernes 
      140       -50        20      -120       240

Y luego hacer la aritmética

In [17]:
140-50+20-120+240

[1] 230

Pero, seamos honesto, eso es muy lento y muy propenso a errores, imagínate que tuvieras un vector con miles de números. ¡Qué flojera! Pero, afortunadamente, R cuenta con una función llamada `sum()` que nos ayuda a hacer la sumatoria de todos los elementos de un vector numérico de forma muy práctica

In [18]:
sum(poker)

[1] 230

Ahora que ya sabes esto, quiero que hagas los siguiente:<br>
- Calcula el total ganado en poker y guárdalo como **total_poker**
- Haz lo mismo para ruleta y guárdalo como **total_ruleta**
- Luego haz el cálculo de lo ganado en total durante la semana, tanto en poker como en ruleta.

¿Qué valor te da? ¿Ganaste o perdiste dinero?

Bien, ¿y si quisieramos comparar en cual juego nos fue mejor? <br>
Para eso podemos usar los símbolos:<br>
- mayor que ( < )
- menor que ( > )
- igual que ( == )
- distinto que ( != ) <br>

Podemos preguntarle a R, por ejemplo, ¿Ganamos lo mismo en poker que en ruleta?

In [19]:
total_poker <-sum(poker)
total_ruleta <-sum(ruleta)

In [20]:
total_poker == total_ruleta

[1] FALSE

Entonces, de ser así probemos a ver si ganamos más en ruleta que en poker

In [21]:
total_poker < total_ruleta

[1] FALSE

OK, veamos el caso contrario

In [22]:
total_poker > total_ruleta

[1] TRUE

Como vemos, con unas sencillas operaciones de lógica podemos determinar si un valor es identico, distinto, mayor o menor que otro.
¡Genial! ahora, ¿qué harías si quisieras seleccionar un valor específico de un vector? <br>
Bueno, para eso podemos usar la siguiente notación:<br>
                       
      nombredelvector[número o nombre de la posición]
                    
Veamos un ejemplo        

In [23]:
poker[3]
poker["Jueves"]

Miércoles 
       20

Jueves 
  -120

Como ves, es muy fácil seleccionar un elemento de un vector. En caso de querer hacer la selección de múltiples elementos, puedes dentro de los corchetes poner otro vector de número con las posiciones que quieras seleccionar. Por ejemplo, elijamos el lunes, miercoles y viernes

In [24]:
poker[c(1,3,5)]

Lunes Miércoles   Viernes 
      140        20       240

De quererlo hacer con número consecutivos puede usar la notación:<br>

       nombre del vector[numeroinicial:numerofinal]
                 
Veamos, elijamos de martes a jueves

In [25]:
poker[2:4]

Martes Miércoles    Jueves 
      -50        20      -120

¡Excelente! Como ves es muy sencillo, ahora me gustaría que hicieras el siguiente ejercicio:<br>
- Con el vector **ruleta** selecciona los valores de Lunes, Martes y Jueves y guardarlos en un vector llamado **LMJ** e imprimelos
- Con el vector poker selecciona Lunes, Martes y Miércoles y guárdalos como **LMM*
- Calcula la media del vector **LMM** 

Para calcular la media existe una función llamada `mean()`, se usa tal y como se usa `sum()`

Ahora continuemos con algo interesante, ¿Qué tal si quisiera saber en qué dias gane y en cuales perdí? Bueno, para determinar eso necesitamos saber si lo que ganamos en cierto día es más que 0, entonces podemos hacer esa operación en R de la siguiente forma

In [26]:
poker > 0

Lunes    Martes Miércoles    Jueves   Viernes 
     TRUE     FALSE      TRUE     FALSE      TRUE

¡Sencillito! Entonces podemos hacer esas mismas operaciones, para saber los días, por ejemplo, en que perdimos :( <br>
Inténtalo


Ahora probemos guardar el resultado de los días en que perdimos en el poker en un vector llamado **perdidas** y seleccionemos usando la notación que ya conocemos para seleccionar elementos de vectores, solo que en lugar de un número o el nombre de un encabezado, coloquemos el nombre del nuevo vector. ¿Qué resultado esperarías?

In [27]:
perdidas <- poker < 0
poker[perdidas]

Martes Jueves 
   -50   -120

### Matrices

Ahora veremos otro tipo de objeto utilizado en R, que consiste de una colección de elementos del mismo tipo de dato (númerico, carácter y lógicos) acomodados en un arreglo fijo de columnas y filas, o 2 dimensiones. Para construir una matriz en R usamos la función `matrix()` <br>
A continuación un ejemplo

In [28]:
matrix(1:9, byrow = TRUE, nrow = 3)


[,1] [,2] [,3]
[1,] 1    2    3   
[2,] 4    5    6   
[3,] 7    8    9

Para prácticar con la construcción y manipulación de matrices, hagamos el siguiente ejercicio con los números de las ganacias en taquilla de las películas de Star Wars en Estados Unidos y fuera de Estados Unidos. Estos están guardados en los siguientes vectores

In [29]:
new_hope <- c(460.998, 314.4)
empire_strikes <- c(290.475, 247.900)
return_jedi <- c(309.306, 165.8)

Guarda esos 3 vectores en una variable llamada **box_office**. <br>
Ahora construye una matriz de 3 filas con cada una de las películas y llámala **starwars_matrix**

In [30]:
box_office <- c(new_hope, empire_strikes, return_jedi)
starwars_matrix <- matrix(box_office, nrow= 3, byrow = TRUE)

In [31]:
box_office

[1] 460.998 314.400 290.475 247.900 309.306 165.800

In [32]:
starwars_matrix

[,1]    [,2] 
[1,] 460.998 314.4
[2,] 290.475 247.9
[3,] 309.306 165.8

Con la función `colnames()` podemos extraer los nombres de las columnas de una matriz y con la función `rownames()` podemos ver los nombres de filas, o como vimos en el caso de los vectores, podemos añadirle nombres.<br> Añade los títulos de filas y columnas con los siguientes vectores:

In [33]:
region <- c("US", "non-US")
titles <- c("A New Hope", "The Empire Strikes Back", "Return of the Jedi")

Ahora imprime la matriz, debe verse así:

                                 US  non-US
    A New Hope              460.998  314.4
    The Empire Strikes Back 290.475  247.9
    Return of the Jedi      309.306  165.8
    

In [34]:
rownames(starwars_matrix) <- titles
colnames(starwars_matrix) <- region
starwars_matrix

US      non-US
A New Hope              460.998 314.4 
The Empire Strikes Back 290.475 247.9 
Return of the Jedi      309.306 165.8

También podemos hacer la suma de los valores de las filas con la función `rowSums()` probemos la función

In [35]:
rowSums(starwars_matrix)

A New Hope The Empire Strikes Back      Return of the Jedi 
                775.398                 538.375                 475.106

Guardemos los resultados en una variable llamada **totales**

In [36]:
totales <- rowSums(starwars_matrix)

Exploremos ahora una función llamada `cbind()` que conjuga matrices o vectores juntos por columnas. <br> Probemosla

In [37]:
cbind(starwars_matrix, totales)

US      non-US totales
A New Hope              460.998 314.4  775.398
The Empire Strikes Back 290.475 247.9  538.375
Return of the Jedi      309.306 165.8  475.106

Guardemos esa matriz como **starwars_total**. <br> Ahora para que práctiquemos, construye la siguiente matriz y guardalo como **starwars_matrix2**:

                            US non-US
    The Phantom Menace   474.5  552.5
    Attack of the Clones 310.7  338.7
    Revenge of the Sith  380.3  468.5

In [38]:
The_Phantom_Menace <- c(474.500, 552.500)
Attack_of_the_clones <- c(310.700, 338.700)
Revenge_of_the_Sith  <-  c(380.300, 468.500)
taquilla <- c(The_Phantom_Menace, Attack_of_the_clones, Revenge_of_the_Sith)
titles2 <- c("The Phantom Menace", "Attack of the Clones", "Revenge of the Sith")
starwars_matrix2 <- matrix(taquilla, nrow = 3, byrow = TRUE)
colnames(starwars_matrix2) <- region
rownames(starwars_matrix2) <- titles2
starwars_matrix2

US    non-US
The Phantom Menace   474.5 552.5 
Attack of the Clones 310.7 338.7 
Revenge of the Sith  380.3 468.5

Si quisieramos unir las filas de uno con el otro, podemos usar la función `rbind(matriz1, matriz2, ...)` 

In [39]:
rbind(starwars_matrix, starwars_matrix2)

US      non-US
A New Hope              460.998 314.4 
The Empire Strikes Back 290.475 247.9 
Return of the Jedi      309.306 165.8 
The Phantom Menace      474.500 552.5 
Attack of the Clones    310.700 338.7 
Revenge of the Sith     380.300 468.5

Guardemos esa matrix como **starwars_todas**

In [40]:
starwars_todas <- rbind(starwars_matrix, starwars_matrix2)

Y así como existe una función para sumar filas, también podemos sumar columnas con la función `colSums()`.<br> Probemosla con nuestra nueva matriz

In [41]:
colSums(starwars_todas)


US   non-US 
2226.279 2087.800

Ahora pongámos a prueba lo aprendido con la construcción de una matriz como la siguiente: <br>

                                US   non-US  totales
    A New Hope               460.998  314.400  775.398
    The Empire Strikes Back  290.475  247.919  538.394
    Return of the Jedi       309.306  165.800  475.106
    The Phantom Menace       474.500  552.500 1027.000
    Attack of the Clones     310.700  338.700  649.400
    Revenge of the Sith      380.300  468.500  848.800
    totales_globales         2226.279 2087.819 4314.098

Y guardalo como **starwars_totales**


In [42]:
totales_globales <- colSums(starwars_todas)
starwars_globales <- rbind(starwars_todas, totales_globales)
totales <- rowSums(starwars_globales)
starwars_totales <- cbind(starwars_globales, totales)
starwars_totales


US       non-US totales 
A New Hope               460.998  314.4  775.398
The Empire Strikes Back  290.475  247.9  538.375
Return of the Jedi       309.306  165.8  475.106
The Phantom Menace       474.500  552.5 1027.000
Attack of the Clones     310.700  338.7  649.400
Revenge of the Sith      380.300  468.5  848.800
totales_globales        2226.279 2087.8 4314.079

¡Genial! Ahora, ¿cómo haríamos la selección de un elemento dentro de una matriz? Bien, pues de una forma muy similar a la de los vectores, con el nombres del objeto seguido de corchetes o brackets `[]`pero al ser una matriz un objeto de 2 dimensiones debemos seleccionar de la siguiente forma:<br> 

            NombreMatriz[NúmeroFila,NúmeroColuma]
            
En caso de querer seleccionar todas las filas o todas la columnas, dejamos el espacio en blanco, ya sea antes o después de la coma, respectivamente. <br>

Probemoslo seleccionando las ganancias de todas la peliculas fuera de los US.

In [43]:
starwars_todas[ ,2]

A New Hope The Empire Strikes Back      Return of the Jedi 
                  314.4                   247.9                   165.8 
     The Phantom Menace    Attack of the Clones     Revenge of the Sith 
                  552.5                   338.7                   468.5

Podemos así, hacer cálculos con esos valores, por ejemplo, podemos cálcular la media o el total

In [44]:
mean(starwars_todas[ ,2])
sum(starwars_todas[ ,2])

[1] 347.9667

[1] 2087.8

También podemos hacer aritmética con estas matrices, por ejemplo, imaginemos que el boleto vale en promedio 5 dolares por persona, entonces podemos dividir la ganancia entre el costo y obtener el número de visitantes

In [45]:
starwars_todas / 5

US      non-US
A New Hope              92.1996  62.88
The Empire Strikes Back 58.0950  49.58
Return of the Jedi      61.8612  33.16
The Phantom Menace      94.9000 110.50
Attack of the Clones    62.1400  67.74
Revenge of the Sith     76.0600  93.70

Ahora tú calcula la média de visitante en estados unidos

In [46]:
visitantes <- starwars_todas / 5
mean(visitantes)

[1] 71.90132

### Factores

El término factor se refiere a un tipo de dato estadísticos utilizado para almacenar variables categóricas. La diferencia entre una variable categórica y una variable continua es que una variable categórica puede pertenecer a un número limitado de categorías. Una variable continua, por otro lado, puede corresponder a un número infinito de valores.<br>

Es importante que R sepa si se trata de una variable continua o categórica, ya que los modelos estadísticos que desarrollará en el futuro tratan a ambos tipos de manera diferente.<br>
 
Un buen ejemplo de una variable categórica es el sexo. En muchas circunstancias, puede limitar las categorías de sexo a "Masculino" o "Femenino". (A veces puede necesitar diferentes categorías. Por ejemplo, es posible que deba considerar la variación cromosómica, los animales hermafroditas o las diferentes normas culturales, pero siempre tendrá un número finito de categorías).

Para convertir una variable a factor usamos la función factor de la siguiente forma:<br>

                    factor(NombreVariable) 
                    
 Experimentemos con los siguientes vector


In [47]:
sexo <- c("Hombre","Mujer","Hombre","Mujer","Mujer","Hombre")
animales <- c("Elefante","Jirafa","Burro","Caballo")
temperaturas <- c("Alta","Baja","Alta","Media","Baja")

In [48]:
factor(sexo)
factor(animales)
factor(temperaturas)

[1] Hombre Mujer  Hombre Mujer  Mujer  Hombre
Levels: Hombre Mujer

[1] Elefante Jirafa   Burro    Caballo 
Levels: Burro Caballo Elefante Jirafa

[1] Alta  Baja  Alta  Media Baja 
Levels: Alta Baja Media

Si quisieramos ver únicamente los Levels de nuestro factor, podemos hacerlo muy facil con la función `levels()` de la siguiente forma:

                   levels(NombreDelFactor)

In [49]:
animales <- factor(animales)
levels(animales)

[1] "Burro"    "Caballo"  "Elefante" "Jirafa"

Pero algunas veces, categorías como el sexo van a estar codificados como M para mujer Y H para hombre.

In [50]:
sexo <- c("H", "M", "M", "H", "M")

In [51]:
factor_sexo <- factor(sexo)
factor_sexo

[1] H M M H M
Levels: H M

Pero si quisieramos modificar eso podemos hacer lo siguiente:

In [52]:
levels(factor_sexo) <- c("Hombre","Mujer")
factor_sexo

[1] Hombre Mujer  Mujer  Hombre Mujer 
Levels: Hombre Mujer

Igual podemos hacer hacer de forma muy sencilla resúmenes con la función `summary()` de la siguiente manera: 

               summary(NombreDeLaVariable)
 
 Por ejemplo:

In [53]:
summary(factor_sexo)
summary(temperaturas)

Hombre  Mujer 
     2      3

   Length     Class      Mode 
        5 character character 

Ahora prueba haciendo la selección del vector factor_sexo y guarda cada uno de los niveles como variables **Hombre** y **Mujer** 

In [54]:
Hombre <- factor_sexo[1]
Mujer <- factor_sexo[2]

¿Qué pasa si los comparamos?

In [55]:
Hombre < Mujer

Warning message in Ops.factor(Hombre, Mujer):
“‘<’ not meaningful for factors”


[1] NA

Como podrás ver, no se pueden hacer comparaciones con factores, así que debes procurar tener muy en cuenta si los valores con los que estás tratando son factores o no. <br>

Ahora práctiquemos, crea un vector llamada **rendimiento** para almacenar las pruebas a 5 analistas, cuyos resultados  nos dicen los rendimientos con la que trabajan.

- Analista 1 es medio
- Analista 2 es bajo
- Analista 3 es bajo
- Analista 4 es medio
- Analista 5 es alto

No es necesario, aún, que especifiques que es un factor

In [56]:
rendimiento <- c("medio","bajo","bajo","medio","alto")

Ahora convierte la variable a factor y guardala como **factor_rendimiento**, imprímela y saca un resumen

In [57]:
factor_rendimiento <- factor(rendimiento)
factor_rendimiento
summary(factor_rendimiento)

[1] medio bajo  bajo  medio alto 
Levels: alto bajo medio

alto  bajo medio 
    1     2     2

¿Qué vemos en los niveles?<br> 
<br> 
<br> 
<br> 



¿Qué si quisieramos cambiar el orden de los nivel? <br> 
Por fortuna la función `factor()`nos admite más de un argumento (los argumentos son las cosas con las que computa una función) Veamos los argumentos que admite `factor()` poniendo ?factor.

In [58]:
?factor

In [59]:
factor_rendimiento <- factor(rendimiento,
                       ordered = TRUE,
                       levels = c("bajo", "medio", "alto"))
factor_rendimiento

[1] medio bajo  bajo  medio alto 
Levels: bajo < medio < alto

In [60]:
summary(factor_rendimiento)

bajo medio  alto 
    2     2     1

Algo importante de tener nuestros factores ordenados es que nos permite hacer comparaciones por ejemplo, respóndeme usando R lo siguiente. 
Es el analista 4 mejor que el analista 2?

In [61]:
analista2 <- factor_rendimiento[2]
analista4 <- factor_rendimiento[4]
analista2 > analista4

[1] FALSE


### Data Frames 

Puedes recordar cuando hablamos sobre matrices que todos los elementos que se colocan en una matriz deben ser del mismo tipo. Entonces, el conjunto de datos en Star Wars solo contenía elementos numéricos. Pero al hacer un estudio se hacen diferentes preguntas que pueden tener diferentes tipos de valores.<br>
En un estudio de mercado, por ejemplo, a menudo tienes preguntas como:

- '¿Estás casado?' o preguntas 'sí / no' (lógico)
- '¿Cuantos años tienes?' (numérico)
- ¿Cuál es su opinión sobre este producto? u otras preguntas 'abiertas' (personaje)

El resultado, es decir, las respuestas de los encuestados a las preguntas formuladas anteriormente, es un conjunto de datos de diferentes tipos de datos. A menudo se encontrará trabajando con conjuntos de datos que contienen diferentes tipos de datos en lugar de solo uno.

Un data frame tiene las variables de un conjunto de datos como columnas y las observaciones como filas. Este será un concepto familiar para quienes provengan de diferentes paquetes de software estadístico como SAS o SPSS.

Un ejemplo de un data frame lo podemos ver al teclear y ejecutar `mtcars` 

In [62]:
mtcars

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

Trabajar con grandes conjuntos de datos no es raro en el análisis de datos. Cuando se trabaja con conjuntos de datos y data frames(extremadamente) grandes, su primera tarea como analista de datos es desarrollar una comprensión clara de su estructura y elementos principales. Por lo tanto, a menudo es útil mostrar solo una pequeña parte de todo el conjunto de datos.

Entonces, ¿cómo hacer esto en R? Bueno, la función `head()` le permite mostrar las primeras observaciones de un marco de datos. Del mismo modo, la función `tail()` imprime las últimas observaciones en su conjunto de datos.

Tanto `head()` como `tail()` imprimen una línea superior llamada 'encabezado', que contiene los nombres de las diferentes variables en su conjunto de datos.

Probemos estas funciones con el data frame mtcars

In [63]:
head(mtcars)
tail(mtcars)

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160  110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160  110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108   93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

mpg  cyl disp  hp  drat wt    qsec vs am gear carb
Porsche 914-2  26.0 4   120.3  91 4.43 2.140 16.7 0  1  5    2   
Lotus Europa   30.4 4    95.1 113 3.77 1.513 16.9 1  1  5    2   
Ford Pantera L 15.8 8   351.0 264 4.22 3.170 14.5 0  1  5    4   
Ferrari Dino   19.7 6   145.0 175 3.62 2.770 15.5 0  1  5    6   
Maserati Bora  15.0 8   301.0 335 3.54 3.570 14.6 0  1  5    8   
Volvo 142E     21.4 4   121.0 109 4.11 2.780 18.6 1  1  4    2

Como podrás ver nos devuelve las primeras y últimas 6 observaciones del data frame, esto lo podemos modificar. por ejemplo, si quisieramos ver los primeros 10 podemos agregar un segundo argumento con el número específico de filas a desplegar

In [64]:
head(mtcars,10)

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360        14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D         24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230          22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280          19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4

Otro método que se usa a menudo para obtener una visión general rápida de sus datos es la función `str()`. La función `str()` le muestra la estructura de su conjunto de datos. Para un data frame te dice:

- El número total de observaciones (por ejemplo, 32 tipos de automóviles)
- El número total de variables (por ejemplo, 11 características del automóvil)
- Una lista completa de los nombres de las variables (por ejemplo, mpg, cyl, etc...)
- El tipo de datos de cada variable (por ejemplo, num)
- Las primeras observaciones

La aplicación de la función `str()` suele ser lo primero que hace al recibir un nuevo conjunto de datos o data frame. Es una excelente manera de obtener más información sobre su conjunto de datos antes de sumergirse en el análisis real.<br>

Probemos esta función

In [65]:
str(mtcars)

'data.frame':	32 obs. of  11 variables:
 $ mpg : num  21 21 22.8 21.4 18.7 18.1 14.3 24.4 22.8 19.2 ...
 $ cyl : num  6 6 4 6 8 6 8 4 4 6 ...
 $ disp: num  160 160 108 258 360 ...
 $ hp  : num  110 110 93 110 175 105 245 62 95 123 ...
 $ drat: num  3.9 3.9 3.85 3.08 3.15 2.76 3.21 3.69 3.92 3.92 ...
 $ wt  : num  2.62 2.88 2.32 3.21 3.44 ...
 $ qsec: num  16.5 17 18.6 19.4 17 ...
 $ vs  : num  0 0 1 1 0 1 0 1 1 1 ...
 $ am  : num  1 1 1 0 0 0 0 0 0 0 ...
 $ gear: num  4 4 4 3 3 3 3 4 4 4 ...
 $ carb: num  4 4 1 1 2 1 4 2 2 4 ...


Dado que el uso de conjuntos de datos integrados no es ni siquiera la mitad de la diversión de crear tus propios conjuntos de datos, hablaremos de como desarrollar conjunto de datos personalmente.<br>

Como primer objetivo, construyamos un data frame que describa las características principales de ocho planetas en nuestro sistema solar. Las características principales de un planeta son:

- El tipo de planeta (terrestre o gaseoso).
- El diámetro del planeta en relación con el diámetro de la   Tierra.
- La rotación del planeta a través del sol en relación con   la de la Tierra.
- Si el planeta tiene anillos o no (VERDADERO o FALSO).

Después de hacer una investigación rápida en Wikipedia https://en.wikipedia.org/wiki/Planet, se pueden crear los vectores necesarios: nombre, tipo, diámetro, rotación y anillos; Estos vectores ya los he codificado para ahorrar tiempo. El primer elemento en cada uno de estos vectores corresponde a la primera observación.

Construye un data frame con la función `data.frame()`. Como argumentos, pasa los vectores de antes: se convertirán en las diferentes columnas de su marco de datos. Debido a que cada columna tiene la misma longitud, los vectores que pasa también deben tener la misma longitud. Pero no olvides que es posible (y probable) que contengan diferentes tipos de datos. Guarda el data frame en una variable llamada **planetas** e imprímelo

In [66]:
nombre <- c("Mercurio", "Venus", "Tierra", "Marte", "Jupiter", "Saturno", "Urano", "Neptuno")
tipo <- c("Terrestre", "Terrestre", "Terrestre", 
          "Terrestre", "Gaseoso", "Gaseoso", "Gaseoso", "Gaseoso")
diametro <- c(0.382, 0.949, 1, 0.532, 11.209, 9.449, 4.007, 3.883)
rotacion <- c(58.64, -243.02, 1, 1.03, 0.41, 0.43, -0.72, 0.67)
anillos <- c(FALSE, FALSE, FALSE, FALSE, TRUE, TRUE, TRUE, TRUE)


In [67]:
planetas <- data.frame(nombre,tipo,diametro,rotacion,anillos)
planetas

nombre   tipo      diametro rotacion anillos
1 Mercurio Terrestre  0.382     58.64  FALSE  
2 Venus    Terrestre  0.949   -243.02  FALSE  
3 Tierra   Terrestre  1.000      1.00  FALSE  
4 Marte    Terrestre  0.532      1.03  FALSE  
5 Jupiter  Gaseoso   11.209      0.41   TRUE  
6 Saturno  Gaseoso    9.449      0.43   TRUE  
7 Urano    Gaseoso    4.007     -0.72   TRUE  
8 Neptuno  Gaseoso    3.883      0.67   TRUE

Checa su estructura

In [68]:
str(planetas)

'data.frame':	8 obs. of  5 variables:
 $ nombre  : Factor w/ 8 levels "Jupiter","Marte",..: 3 8 6 2 1 5 7 4
 $ tipo    : Factor w/ 2 levels "Gaseoso","Terrestre": 2 2 2 2 1 1 1 1
 $ diametro: num  0.382 0.949 1 0.532 11.209 ...
 $ rotacion: num  58.64 -243.02 1 1.03 0.41 ...
 $ anillos : logi  FALSE FALSE FALSE FALSE TRUE TRUE ...


Similar a los vectores y las matrices, selecciona elementos de un data frame con la ayuda de corchetes []. Al usar una coma, puede indicar qué seleccionar de las filas y las columnas, respectivamente. Por ejemplo:

- my_df[1,2] selecciona el valor en la primera fila y la segunda columna en my_df.
- my_df[1: 3,2: 4] selecciona las filas 1, 2, 3 y las columnas 2, 3, 4 en my_df.

A veces desea seleccionar todos los elementos de una fila o columna. Por ejemplo, my_df[1,] selecciona todos los elementos de la primera fila. Ahora apliquemos esta técnica en planetas

In [69]:
# Imprime el diámetro de Mercurio (fila 1, columna 3)
planetas[1,3]
# Imprime los datos para Marte (toda la cuarta fila)
planetas[4,]

[1] 0.382

nombre tipo      diametro rotacion anillos
4 Marte  Terrestre 0.532    1.03     FALSE

En lugar de usar números para seleccionar elementos de un data frame, también puede usar los nombres de las variables para seleccionar columnas de un marco de datos.

Suponga que desea seleccionar los primeros tres elementos de la columna 'tipo'. Una forma de hacer esto es:

                  planetas_df [1: 3,2]
                    
Una posible desventaja de este enfoque es que debe conocer (o buscar) el número de columna 'tipo', lo que se vuelve difícil si tiene muchas variables. A menudo es más fácil usar el nombre de la variable:

                planets_df [1: 3, "tipo"]
Práctiquemos

In [70]:
# Selecciona los primeros 5 valores de la columna 
# de diámetro
planetas[1:5, "diametro"]

[1]  0.382  0.949  1.000  0.532 11.209

A menudo querrás seleccionar una columna completa, es decir, una variable específica de un data frame. Si desea seleccionar todos los elementos de la variable 'diámetro', por ejemplo, ambos harán el truco:

                        planetas[, 3]
                              ó
                   planetas[, "diámetro"]
                   
Sin embargo, hay un atajo. Si sus columnas tienen nombres, puede usar el signo $:

                     planetas$diámetro
                     
Probemoslo

In [71]:
# Selecciona la variable anillos de planetas y guardala como
# anillos


# Imprime la variable

Probablemente recuerdes de la escuela secundaria que algunos planetas de nuestro sistema solar tienen anillos y otros no. Lamentablemente no puedes recordar sus nombres. ¿Podría R ayudarnos?

Si escribe anillos en la consola, obtendrás:

         [1] FALSE FALSE FALSE FALSE TRUE TRUE                          TRUE TRUE
         
Esto significa que las primeras cuatro observaciones (o planetas) no tienen un anillo (FALSE), pero las otras cuatro sí (TRUE). Sin embargo, no obtienes una buena visión general de los nombres de estos planetas, su diámetro, etc. Tratemos de usar el vector **anillos** para seleccionar los datos de los cuatro planetas con anillos.

In [72]:
# selecciona todas las columnas para planetas con anillos
planeta[anillos, ]

ERROR: Error in eval(expr, envir, enclos): objeto 'planeta' no encontrado


Entonces, ¿qué aprendiste exactamente de este ejercicio? Seleccionaste un subconjunto de un data frame (planetas) en función de si cierta condición era verdadera (anillos o no anillos) y logró extraer todos los datos relevantes.<br>

Ahora, avancemos un nivel y usemos la función `subset()`. Deberias ver que la función `subset()`funciona como un atajo para hacer exactamente lo mismo que hizo en los ejercicios anteriores.

        subset(my_df, subset = alguna_condición)
        
El primer argumento de `subset()` especifica el conjunto de datos para el que desea un subconjunto. Al agregar el segundo argumento, le da a R la información y las condiciones necesarias para seleccionar el subconjunto correcto.

El siguiente código le dará exactamente el mismo resultado que obtuvo en el ejercicio anterior, pero esta vez, ¡no necesitamos el vector **anillos**!

In [73]:
subset(planetas, subset = anillos)

nombre  tipo    diametro rotacion anillos
5 Jupiter Gaseoso 11.209    0.41    TRUE   
6 Saturno Gaseoso  9.449    0.43    TRUE   
7 Urano   Gaseoso  4.007   -0.72    TRUE   
8 Neptuno Gaseoso  3.883    0.67    TRUE

Ahora un ejercicio, usa `subset()` en **planetas** para seleccionar planetas que tengan un diámetro menor que la Tierra. Debido a que la variable de diámetro es una medida relativa del diámetro del planeta Tierra, su condición es diámetro < 1.

In [74]:
subset(planeta, subset = diametro < 1)

ERROR: Error in subset(planeta, subset = diametro < 1): objeto 'planeta' no encontrado


Hacer y crear clasificaciones es uno de los asuntos favoritos de la humanidad. Estas clasificaciones pueden ser útiles (las mejores universidades del mundo), entretenidas (las estrellas de cine más influyentes) o inútiles (el mejor cosplay del 2020).

En el análisis de datos, puedes ordenar sus datos de acuerdo con una determinada variable en el conjunto de datos. En R, esto se hace con la ayuda de la función `order()`.

`order()` es una función que le brinda la posición clasificada de cada elemento cuando se aplica a una variable, como un vector, por ejemplo:

In [75]:
a <- c (100, 10, 1000)
order(a) 

[1] 2 1 3

10, que es el segundo elemento en a, es el elemento más pequeño, por lo que 2 viene primero en el resultado de `order(a)`. 100, que es el primer elemento en a, es el segundo elemento más pequeño, por lo que 1 ocupa el segundo lugar en la salida del `order(a)`.
Esto significa que podemos usar el resultado de `order(a)` para reorganizar:

In [76]:
a[order(a)]

[1]   10  100 1000

Bien, ahora que comprendes la función `order()`, hagamos algo útil con ella. Imaginemos que necesitamos reorganizar tu data frame de modo que comience con el planeta más pequeño y termine con el más grande. Un reordenamiento de la columna 'diámetro'.
¿Cómo le harías?

In [77]:
# Usa order() para crear las posiciones
posiciones <- order(planetas$diametro) 

# Usa posiciones para acomodar los planetas por diámetro
planetas[posiciones, ]

nombre   tipo      diametro rotacion anillos
1 Mercurio Terrestre  0.382     58.64  FALSE  
4 Marte    Terrestre  0.532      1.03  FALSE  
2 Venus    Terrestre  0.949   -243.02  FALSE  
3 Tierra   Terrestre  1.000      1.00  FALSE  
8 Neptuno  Gaseoso    3.883      0.67   TRUE  
7 Urano    Gaseoso    4.007     -0.72   TRUE  
6 Saturno  Gaseoso    9.449      0.43   TRUE  
5 Jupiter  Gaseoso   11.209      0.41   TRUE

En este punto del curso ya está familiarizado con:

- Vectores (matriz unidimensional): puede contener valores numéricos, de caracteres o lógicos. Todos los elementos en un vector tienen el mismo tipo de datos.
- Matrices (matriz bidimensional): puede contener valores numéricos, de caracteres o lógicos. Todos los elementos en una matriz tienen el mismo tipo de datos.
- Data frames (objetos bidimensionales): pueden contener valores numéricos, de caracteres o lógicos. Dentro de una columna, todos los elementos tienen el mismo tipo de datos, pero diferentes columnas pueden ser de diferentes tipos de datos.

### Listas
Una lista en R es similar a su lista de tareas pendientes en el trabajo o la escuela: los diferentes elementos en esa lista probablemente tienen diferente duración, características y tipo de actividad que debe realizarse.

Una lista en R le permite reunir una variedad de objetos bajo un nombre (es decir, el nombre de la lista) de forma ordenada. Estos objetos pueden ser matrices, vectores, marcos de datos, incluso otras listas, etc. Ni siquiera se requiere que estos objetos estén relacionados entre sí de ninguna manera.¡Puedes almacenar prácticamente cualquier información en ella!<br>

¡Creemos nuestra primera lista! Para construir una lista, use la función `list()`:

           mi_lista <- list(componente1, componente2 ...)

Construye una lista, llamada **lista**, que contenga las variables **vector**, **matriz** y **df** como componentes de la lista.
           



In [78]:
vector <- 1:10 
matriz <- matrix(1:9, ncol = 3)
df <- mtcars[1:10,]

# Construye la lista con estos diferentes elementos:
lista<- list(vector, matrix, df)

In [79]:
lista

[[1]]
 [1]  1  2  3  4  5  6  7  8  9 10

[[2]]
function (data = NA, nrow = 1, ncol = 1, byrow = FALSE, dimnames = NULL) 
{
    if (is.object(data) || !is.atomic(data)) 
        data <- as.vector(data)
    .Internal(matrix(data, nrow, ncol, byrow, dimnames, missing(nrow), 
        missing(ncol)))
}
<bytecode: 0x7f8845927858>
<environment: namespace:base>

[[3]]
                   mpg cyl  disp  hp drat    wt  qsec vs am gear carb
Mazda RX4         21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag     21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4
Datsun 710        22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive    21.4   6 258.0 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout 18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
Valiant           18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1
Duster 360        14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4
Merc 240D         24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2
Merc 230          22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2
Merc 280          19.2   6 167.6 123 3.92 3.440 18.30  1  0    4    4

Al igual que en su lista de tareas pendientes, debe evitar no saber o recordar qué significan los componentes de su lista. Es por eso que debes darles nombres:

            lista <- list(nombre1 = comp1,
                          nombre2 = comp2)

Esto crea una lista con componentes que se denominan nombre1, nombre2, etc. Si desea nombrar sus listas después de haberlas creado, puede usar la función `names()` como lo hiciste con los vectores. Los siguientes comandos son totalmente equivalentes a la asignación anterior:

                lista <- list(comp1, comp2)
           names(lista) <- c ("nombre1", "nombre2")

Ahora añade a tu lista anterior los nombres **vector**, **matriz** y **df** e imprímela

In [80]:
names(lista) <- c("vector","matriz", "df")
lista

$vector
 [1]  1  2  3  4  5  6  7  8  9 10

$matriz
function (data = NA, nrow = 1, ncol = 1, byrow = FALSE, dimnames = NULL) 
{
    if (is.object(data) || !is.atomic(data)) 
        data <- as.vector(data)
    .Internal(matrix(data, nrow, ncol, byrow, dimnames, missing(nrow), 
        missing(ncol)))
}
<bytecode: 0x7f8845927858>
<environment: namespace:base>

$df
                   mpg cyl  disp  hp drat    wt  qsec vs am gear carb
Mazda RX4         21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag     21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4
Datsun 710        22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive    21.4   6 258.0 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout 18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
Valiant           18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1
Duster 360        14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4
Merc 240D         24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2
Merc 230          22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2
Merc 280          19.2   6 167.6 123 3.92 3.440 18.30  1  0    4    4

Para repasar lo que hemos visto hasta ahora quiero que armes los siguientes objetos:

- una variable llamada **peli** con el nombre de El resplando
- Un vector llamado **act** con los siguientes nombres: 
        Jack Nicholson
        Shelley Duvall
        Danny Lloyd
        Scatman Crothers
        Barry Nelson
- Un data frame con 3 columnas: 
                        puntaje
                        fuente
                        comentarios
  y en cada una de ellas los siguientes valores:
  - primera: 4.5, 4.0, 5.0
  - segunda: IMDb1,IMDb2,IMDb3
  - tercera: La mejor película de terror que he visto,
           Una película realmente brillante y aterradora de Stanley Kubrick
           Una obra maestra del horror psicológico




In [81]:
peli <- "El resplandor"
act <- c("Jack Nicholson", "Shelley Duvall","Danny Lloyd","Scatman Crothers","Barry Nelson")
rev <- data.frame(puntaje = c(4.5,4.0,5.0),
                  fuente = c("IMDb1","IMDB2","IMDB3"),
                  comentarios = c("La mejor película de terror que he visto",
                                  "Una película realmente brillante y aterradora de Stanley Kubrick",
                                  "Una obra maestra del horror psicológico"))

peli
act
rev

[1] "El resplandor"

[1] "Jack Nicholson"   "Shelley Duvall"   "Danny Lloyd"      "Scatman Crothers"
[5] "Barry Nelson"

puntaje fuente
1 4.5     IMDb1 
2 4.0     IMDB2 
3 5.0     IMDB3 
  comentarios                                                     
1 La mejor película de terror que he visto                        
2 Una película realmente brillante y aterradora de Stanley Kubrick
3 Una obra maestra del horror psicológico

Ahora construye una lista llamada **el_resplandor** con estos 3 objectos y nombralos:

- película
- actores
- reseñas

In [82]:
el_resplandor <- list(película = peli,
                     actores = act,
                     reseñas = rev)
el_resplandor

$película
[1] "El resplandor"

$actores
[1] "Jack Nicholson"   "Shelley Duvall"   "Danny Lloyd"      "Scatman Crothers"
[5] "Barry Nelson"    

$reseñas
  puntaje fuente
1     4.5  IMDb1
2     4.0  IMDB2
3     5.0  IMDB3
                                                       comentarios
1                         La mejor película de terror que he visto
2 Una película realmente brillante y aterradora de Stanley Kubrick
3                          Una obra maestra del horror psicológico

Las listas a menudo estarán compuestas de numerosos elementos y componentes. Por lo tanto, obtener un solo elemento, múltiples elementos o un componente no siempre es sencillo. <br>
Una forma de seleccionar un componente es usar la posición numerada de ese componente. Por ejemplo, para "agarrar" el primer componente de la lista **el_resplandor** escriba:


In [83]:
el_resplandor[[1]]


[1] "El resplandor"

Para seleccionar elementos de vectores, se utilizan corchetes simples, en este caso usamos corchetes dobles.<br>
También puede referirse a los nombres de los componentes, con [[ ]] o con el signo $. Ambos seleccionarán el data frame que representa las reseñas:

In [84]:
el_resplandor[["reseñas"]]
el_resplandor$reseñas

puntaje fuente
1 4.5     IMDb1 
2 4.0     IMDB2 
3 5.0     IMDB3 
  comentarios                                                     
1 La mejor película de terror que he visto                        
2 Una película realmente brillante y aterradora de Stanley Kubrick
3 Una obra maestra del horror psicológico

puntaje fuente
1 4.5     IMDb1 
2 4.0     IMDB2 
3 5.0     IMDB3 
  comentarios                                                     
1 La mejor película de terror que he visto                        
2 Una película realmente brillante y aterradora de Stanley Kubrick
3 Una obra maestra del horror psicológico

Además de seleccionar componentes, a menudo necesita seleccionar elementos específicos de estos componentes. Por ejemplo:

In [85]:
el_resplandor[[2]][1]

[1] "Jack Nicholson"

Selecciona del segundo componente, **actores** (shining_list [[2]]) y dentro de ese componente el primer elemento ([1]). <br>
Practica seleccionando el vector de actores y de este elige el 3er elemento

In [86]:
el_resplandor$actores
el_resplandor[[2]][3]

[1] "Jack Nicholson"   "Shelley Duvall"   "Danny Lloyd"      "Scatman Crothers"
[5] "Barry Nelson"

[1] "Danny Lloyd"